### PR TITLE
NIFI-7886: FetchAzureBlobStorage, FetchS3Object, and FetchGCSObject processors should be able to fetch ranges

### PR DESCRIPTION
**Pull request by pkelly-nifi**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

Enables pulling objects and blobs by byte ranges for FetchAzureBlobStorage, FetchS3Object, and FetchGCSObject processors as described in NIFI-7886.  Adds RANGE_START and RANGE_LENGTH parameters to all three processors and adjusts the API calls as necessary.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [X] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [X] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [X] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [X] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [X] Have you written or updated unit tests to verify your changes?
- [X] Have you verified that the full build is successful on JDK 8?
- [X] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [X] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by pvillard31** [comment]
Is it also supported for the recently added ADLS Gen2 processor? If yes, would be nice to add it as well.

**Comment by pkelly-nifi** [comment]
Yes, ADLS Gen2 does support pulling by range.  Good suggestion.  I just committed an update for that one as well.

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/FetchS3Object.java
Comment (lucasmoten) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/FetchS3Object.java:174]: If rangeLength is 0, won't this result in a returned range of 1 byte?

Comment (pkelly-nifi) [reply nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/FetchS3Object.java:174]: It doesn't seem to return a range of 1.  If rangeLength is 0, S3 throws an error about it being an invalid range and the file routes to the Failure queue.  A rangeLength of 1B downloads 1 byte as expected.  Am I missing something?  I'll update the validator for rangeLength to use createDataSizeBoundsValidator(1, Long.MAX_VALUE) so that we don't ever get to the zero length error condition.

Comment (lucasmoten) [reply nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/FetchS3Object.java:174]: I'm trying to test from your recent changes on the branch and ran into an issue when I set these property values for FetchS3Object
- Range Start = 1
- Range Length = 1

The processor can't validate with that setting as it produces this validation warning:
- 's3-object-range-start' validated against '1' is invalid because Must be of format <Data Size> <Data Unit> where <Data Size> is a non-negative integer and <Data Unit> is a supported Data Unit, such as: B, KB, MB, GB, TB
- Same validation error for 's3-object-range-length'. 


To clarify my original concern, should require Range Length  be > 0 if Range Start is set to a value.  
Since range requests start at 0 byte, It should be valid to have 
- Range Start = 0
- Range Length = 1
with a result indicating that this is requesting just the first byte from the resource.

Comment (lucasmoten) [reply nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/FetchS3Object.java:174]: As a follow up, if I change
- Range Start = 1B
- Range Length = 1B

What currently gets returned is bytes 1 and 2, which is returning 1 more byte then requested.

Comment (lucasmoten) [reply nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/FetchS3Object.java:174]: At time of writing, im testing this with a small file in an s3 bucket, and with three different FetchS3Object calls on the same, 

See this image of a flow depicting the three different FetchS3Object processors, which only differ in what range start and range length is requested 
Screenshot from 2020-10-22 18-02-52

With the following hex representation of a file that is the source being retrieved ... 
```
0x00000000	5B	2E	53	68	65	6C	6C	43	6C	61	73	73	49	6E	66	6F	[.ShellClassInfo
0x00000010	5D	0D	0A	43	4C	53	49	44	3D	7B	36	34	35	46	46	30	]..CLSID={645FF0
0x00000020	34	30	2D	35	30	38	31	2D	31	30	31	42	2D	39	46	30	40-5081-101B-9F0
0x00000030	38	2D	30	30	41	41	30	30	32	46	39	35	34	45	7D	0D	8-00AA002F954E}.
0x00000040	0A	4C	6F	63	61	6C	69	7A	65	64	52	65	73	6F	75	72	.LocalizedResour
0x00000050	63	65	4E	61	6D	65	3D	40	25	53	79	73	74	65	6D	52	ceName=@%SystemR
0x00000060	6F	6F	74	25	5C	73	79	73	74	65	6D	33	32	5C	73	68	oot%\system32\sh
0x00000070	65	6C	6C	33	32	2E	64	6C	6C	2C	2D	38	39	36	34	0D	ell32.dll,-8964.
0x00000080	0A	.
```

from left to right ...
- start 0, length 0 ... yields 1 byte ... viewing queue in nifi, content in hex mode has   `0x00000000    5B  10    [.`
  - Im not sure whether this should error (returning nothing), or return the entire file?
- start 0, length 1 ... yields 2 bytes .. viewing queue in nifi, content in hex mode has  `0x00000000    5B  2E    [.`
  - I was expecting this to return `5B`
- start 1, length 1 ... yields 2 bytes .. viewing queue in nifi, content in hex mode has  `0x00000000    2E  53    .S`
  - I was expecting this to return `2E`

Comment (lucasmoten) [reply nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/FetchS3Object.java:174]: Maybe change 
```
request.setRange(rangeStart, rangeStart + rangeLength);
```
to
```
request.setRange(rangeStart, rangeStart + rangeLength - 1);
```

Comment (pkelly-nifi) [reply nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/FetchS3Object.java:174]: Thanks for your feedback.  I just tested it again against real AWS S3 and you are absolutely right.  This is apparently a bug in a third party product we've been using.  I will update with the -1 tomorrow.

As for the validation errors you mentioned, this is using DataSize validation rather than pure long values, which allows shortcuts such as Range start: 0B, Range Length: 1GB, followed by Range start: 1GB, Range length: 1GB.  So to use plain bytes, it requires adding a 'B' to the end.  It seemed slightly cleaner than requiring someone to calculate out to the exact byte.  Do you have a strong opinion against this syntax?  Or a suggestion for making it clearer in the documentation?  I'd be happy to update it.

Comment (pkelly-nifi) [reply nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/FetchS3Object.java:174]: I just committed changes with the -1 for the range length calculation and changed the validators to require a length of at least 1B if it is specified.  I don't think range length should be required when range start is specified because sometimes we have a need to read to the end of the file from a starting offset, and reading to the end is the default behavior if range length is not specified.  Please let me know your thoughts.

Comment (lucasmoten) [reply nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/FetchS3Object.java:174]: Personally I'm ok with using 0B, though I'd defer to others when it comes to standards for that.  If its kept that way, I agree, it should have documentation guidance improved.  I plan to take a look at this in the next 24 hours and will report back.

Comment (lucasmoten) [reply nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/FetchS3Object.java:174]: My test results using a file with the following contents as input

```
0x00000000  5B 2E 53 68 65 6C 6C 43 6C 61 73 73 49 6E 66 6F  [.ShellClassInfo
0x00000010  5D 0D 0A 43 4C 53 49 44 3D 7B 36 34 35 46 46 30  ]..CLSID={645FF0
0x00000020  34 30 2D 35 30 38 31 2D 31 30 31 42 2D 39 46 30  40-5081-101B-9F0
0x00000030  38 2D 30 30 41 41 30 30 32 46 39 35 34 45 7D 0D  8-00AA002F954E}.
0x00000040  0A 4C 6F 63 61 6C 69 7A 65 64 52 65 73 6F 75 72  .LocalizedResour
0x00000050  63 65 4E 61 6D 65 3D 40 25 53 79 73 74 65 6D 52  ceName=@%SystemR
0x00000060  6F 6F 74 25 5C 73 79 73 74 65 6D 33 32 5C 73 68  oot%\system32\sh
0x00000070  65 6C 6C 33 32 2E 64 6C 6C 2C 2D 38 39 36 34 0D  ell32.dll,-8964.
0x00000080  0A                                               .
```

| Range Start | Range Length | Result | Notes |
| --- | --- | --- | --- | 
| 0B | 0B | PASS | 0B no longer allowed for length, must be set to empty string/unset, and validation picks this up |
| 0B | 1B | PASS | 1 byte returned, hex viewer shows 5B 10 (ok, see note below) |
| 1B | 1B | PASS | 1 byte returned, hex viewer shows 2E 10 (ok, see note below) |
| 0B | 2B | PASS | 2 bytes returned, hex viewer shows 5B 2E |
| 120B | 120B | PASS | 9 bytes returned, hex viewer showes 6C 2C 2D 38 39 36 34 0D 0A, the final 9 bytes of the file |
| 0B | unset | PASS | 129 bytes returned, hex viewer shows full contents of file |
| 120B | unset | PASS | 9 bytes returned, hex viewer showes 6C 2C 2D 38 39 36 34 0D 0A, the final 9 bytes of the file |

Known Issues
- There is some bug in the hex viewer that shows an extra x10 character when the content is only 1 byte.  This isn't introduced by this capability.

The above testing was done manually, so this is something that could be applied in a unit test to automate

**Review by lucasmoten — APPROVED** [review]
I approve these changes

**Comment by lucasmoten** [comment]
@pvillard31 do you have any other recommendations for this pull request?

**Comment by pvillard31** [comment]
No, I just don't have the time to make any tests on my side with all the cloud providers right now (it seems this has been tested extensively on AWS, but we need to make sure this works well on ADLS/BlobStorage/GCS as well). If any other committer can give a +1, I'm good with this.

**Comment by jfrazee** [comment]
@lucasmoten @pvillard31 I'll review and do some testing on this.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureBlobStorage.java
Comment (jfrazee) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureBlobStorage.java:63]:
```suggestion
            .name("range-start")
```

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureBlobStorage.java
Comment (jfrazee) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureBlobStorage.java:73]:
```suggestion
            .name("range-length")
```

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureDataLakeStorage.java
Comment (jfrazee) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureDataLakeStorage.java:52]:
```suggestion
            .name("range-start")
```

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureDataLakeStorage.java
Comment (jfrazee) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureDataLakeStorage.java:62]:
```suggestion
            .name("range-length")
```

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureDataLakeStorage.java
Comment (jfrazee) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureDataLakeStorage.java:72]:
```suggestion
            .name("number-of-retries")
```

File: nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/storage/FetchGCSObject.java
Comment (jfrazee) [inline nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/storage/FetchGCSObject.java:238]:
```suggestion
            if (rangeLength == null) {
```

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureDataLakeStorage.java
Comment (jfrazee) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureDataLakeStorage.java:118]:
```suggestion
            flowFile = session.write(flowFile, os -> fileClient.readWithResponse(os, fileRange, retryOptions, null, false, null, Context.NONE));
```

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/FetchS3Object.java
Comment (jfrazee) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/FetchS3Object.java:101]:
```suggestion
            .name("range-start")
```

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/FetchS3Object.java
Comment (jfrazee) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/FetchS3Object.java:111]:
```suggestion
            .name("range-length")
```

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/FetchS3Object.java
Comment (jfrazee) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/FetchS3Object.java:174]:
```suggestion
        if (rangeLength != null) {
```

**Review by jfrazee — CHANGES_REQUESTED** [review]
@pkelly-nifi Hey, I looked this over and tested it thoroughly on Azure. The implementation looks good and it's working as expected.

I have some tiny suggestions about making the property names consistent with the pre-existing code, and I think it'd be useful to add tests to the ITs, where they exist. What do you think?

To help with my testing, I went ahead and created some ITs for Blob and ADLS (see [here](https://gist.github.com/jfrazee/33dc4221437f8a94ef66dc56f0d12f2a)). If you end up including them as part of this PR feel free to grab these, if you want to do something different that's cool too.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureBlobStorage.java
Comment (pkelly-nifi) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureBlobStorage.java:102]: Hi @jfrazee,

Thank you for reviewing this PR and for your input.  I very much appreciate it.  I think it might be useful to keep the azure.length attribute as the original remote length since it could be used to verify the final file size after reassembly, if that is what is ultimately desired.  I'm open to suggestions if you disagree.

Comment (pkelly-nifi) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureBlobStorage.java:102]: Thanks also for the ITs.  I will grab yours and include it and try to add some for the other services.  Thanks for your help.

**Comment by pvillard31** [comment]
@jfrazee - are there still requested changes on this pull request?

**Comment by pkelly-nifi** [comment]
> @jfrazee - are there still requested changes on this pull request?

I haven't had a chance yet to add any tests, so we're still waiting on that.  If someone else wants to kick in some tests that'd be much appreciated.

**Comment by jfrazee** [comment]
@pvillard31 @pkelly-nifi The ITs are a nice to have but we can also handle those separately.

The testing on GCS is more important. I'll look into that again today.

**Comment by pkelly-nifi** [comment]
@jfrazee I committed an update that accounts for the scenario you described for FetchGCSObject where start range > object length.  Please let me know your thoughts.  I also included your ITs for Azure.  Thank you.

File: nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/storage/FetchGCSObject.java
Comment (jfrazee) [inline nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/storage/FetchGCSObject.java:237]: Minor thing, in general we try to wrap the debugs in an `isDebugEnabled()`. This one should (hopefully) be reached only very rarely so definitely not a big deal, but thought I'd mention it for the future.
```suggestion
                if (getLogger().isDebugEnabled() {
                    getLogger().debug("Start position: {}, blob size: {}", new Object[] {rangeStart, blob.getSize()});
                }
```

**Comment by jfrazee** [comment]
@pkelly-nifi I plan to spend some time on this tomorrow. Thanks for the updates!

File: nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/storage/FetchGCSObject.java
Comment (pkelly-nifi) [reply nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/storage/FetchGCSObject.java]: Thank you, that makes sense.

**Review by jfrazee — APPROVED** [review]
+1

@pkelly-nifi Thanks for all the work on this!

**Pull request closed without merging**
